**The only difference with the previous version is that the constrains in Dv are released from 1000kms to 500 kms.**

The result extraction is automated. The result table only contains the best run.

In this version, we still distinguish primary galaxies from isolated.
- Primary galaxies are the closest of the LOS (when multiple are present), with a closest neighbour at least at b_sep (not high sthg in 10 - 30 kpc). 

- Isolated galaxies are chosen conservatively: we only conside galaxies alone in 100 kpc and +-1000kms (no concept of satellite galaxies with low mass) and no neighbour closer than B +b_sep (to avoid cases like 2 galaxies at resp 97 and 103 kpc).

In [37]:
import numpy as np
import matplotlib.pyplot as plt
import galpak
#from galpak import DefaultModel, ModelSersic

import os

import numpy as np
import scipy as sc
import matplotlib.pyplot as plt
import pandas as pd
from astropy.table import Table
import astropy.units as u
from astropy.coordinates import SkyCoord
from astropy import constants as const
from astropy import units as u
from astropy.coordinates import SkyCoord, Distance
from astropy.cosmology import Planck15
from astropy.cosmology import WMAP9 as cosmo
from scipy.stats import binned_statistic
import matplotlib.backends.backend_pdf

import mpdaf
from mpdaf.obj import Cube
from mpdaf.drs import PixTable
from galpak import MoffatPointSpreadFunction, GaussianLineSpreadFunction
from mpdaf.sdetect import Source

from astropy.io import fits 

# the home made tool package:
import galpak_tools as gt

%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [51]:
input_path = "/muse/MG2QSO/private/catalogs_dr2/dr2_vbeta/"
output_path = "/muse/MG2QSO/private/analysis/galpak_dr2/"
dr2_path = "/muse/MG2QSO/private/catalogs_dr2/dr2_vbeta/MEGAFLOW_master_dr2.v20230605.fits"
fields_info_path = "/muse/MG2QSO/private/analysis/galpak_dr2/fields_info_ebv.csv"
Abs_path = "/muse/MG2QSO/private/analysis/galpak_dr2/MgII_all_absorption_systems_Sowgat_Dec2020_modified_v2023_matchDR1.csv"
media_path = "/media/mcherrey/foghar3/galpak_dr2/"
primary_tab_path = output_path + "recap_tab.csv"

In [52]:
runs = gt.extract_results_all(output_path, decomp = False)

test_stacking
J2137p0012
J2137p0012_source-31114
     run_OII_multinest
     run_OII_3
J2137p0012_source-31103
     run_OII_multinest
     run_OII_3
J2137p0012_source-31026
J2137p0012_source-31061
     run_OII_3
J2137p0012_source-31017
J2137p0012_source-31062
J2137p0012_source-31010
J2137p0012_source-31109
J2137p0012_source-31113
     run_cont
     run_OII_3
J2137p0012_source-31011
J2137p0012_source-31082
     run_cont
     run_OII_multinest
     run_OII_3
J2137p0012_source-31037
J2137p0012_source-31081
     run_OII_3
J2137p0012_source-31003
     run_HALPHA
     run_OII_3
J2137p0012_source-31043
J2137p0012_source-31023
J2137p0012_source-31056
J2137p0012_source-31099
     run_OII_3
J2137p0012_source-31073
J2137p0012_source-31049
J2137p0012_source-31129
     run_OII_3
J2137p0012_source-31066
J2137p0012_source-31072
J2137p0012_source-31124
     run_OII_3
J2137p0012_source-31032
J2137p0012_source-31018
J2137p0012_source-31015
J2137p0012_source-31130
J2137p0012_source-31046
J2137p0012_sourc

J0134p0051_source-17085
     run_OII_multinest
     run_OII_3
J0134p0051_source-17094
     run_OII_3
J0134p0051_source-17102
     run_cont
     run_OII_multinest
     run_OII_3
J0134p0051_source-17063
     run_cont
     run_HALPHA
J0134p0051_source-17034
     run_OII_multinest
     run_OII_3
J0134p0051_source-17044
     run_OII_multinest
J0134p0051_source-17056
     run_OII_multinest
     run_OII_3
J0134p0051_source-17059
     run_OII_3
J0134p0051_source-17076
     run_OII_multinest
     run_OII_3
J0134p0051_source-17075
     run_OII_multinest
     run_OII_3
J0134p0051_source-17018
J0134p0051_source-17013
J0134p0051_source-17039
     run_OII_multinest
     run_OII_3
J0134p0051_source-17053
     run_cont
     run_OII_3
J0134p0051_source-17035
     run_OII_multinest
     run_OII_3
J0134p0051_source-17026
J0134p0051_source-17011
J0134p0051_source-17060
     run_OII_test_psf
     run_OII_multinest
     run_OII_3
J0134p0051_source-17089
     run_OII_3
J0134p0051_source-17019
J0103p1332
J010

J0015m0751_source-13031
J0015m0751_source-13022
J0015m0751_source-13110
     run_OII_3
J0015m0751_source-13118
J0015m0751_source-13036
J0015m0751_source-13065
J0015m0751_source-13086
     run_OII_multinest
     run_OII_3
J0015m0751_source-13098
     run_OII_3
J0015m0751_source-13117
J0015m0751_source-13111
     run_OII_3
J0015m0751_source-13025
J0015m0751_source-13093
     run_OII_multinest
J0015m0751_source-13092
     run_OII_3
J0015m0751_source-13114
J0015m0751_source-13014
     run_OII_3
J0015m0751_source-13049
J0015m0751_source-13129
     run_cont
     run_OII_multinest
     run_OII_1
     run_OII_3
J0015m0751_source-13061
     run_OII_3
J0015m0751_source-13006
     run_OII_3
J0015m0751_source-13052
J0015m0751_source-13122
J0015m0751_source-13130
     run_cont
     run_OII_3
J0015m0751_source-13078
     run_OII_multinest
     run_OII_3
J0015m0751_source-13015
J0015m0751_source-13002
J0015m0751_source-13054
J0015m0751_source-13019
J0015m0751_source-13021
J0015m0751_source-13103
    

J0838p0257_source-20022
     run_OII_3
J0838p0257_source-20082
J0838p0257_source-20023
J0838p0257_source-20100
     run_OII_3
J0838p0257_source-20070
     run_OII_multinest
     run_OII_3
J0838p0257_source-20099
     run_OII_3
J0838p0257_source-20049
J0838p0257_source-20043
     run_OII_3
J0838p0257_source-20001
J0838p0257_source-20086
     run_OII_3
J0838p0257_source-20057
     run_OII_3
J0838p0257_source-20040
     run_OII_3
J0838p0257_source-20065
     run_OII_3
J0838p0257_source-20069
     run_OII_multinest
     run_OII_3
J0838p0257_source-20081
     run_OII_3
J0838p0257_source-20097
     run_OII_3
J0838p0257_source-20032
J0838p0257_source-20046
     run_OII_3
J0838p0257_source-20089
J0838p0257_source-20061
     run_OII_3
J0838p0257_source-20090
     run_OII_3
J0838p0257_source-20018
J0838p0257_source-20003
J0838p0257_source-20029
J0838p0257_source-20062
     run_OII_multinest
J0838p0257_source-20044
     run_OII_multinest
     run_OII_3
J0838p0257_source-20073
     run_OII_multine

J1509p1506_source-30088
J1509p1506_source-30046
J1509p1506_source-30017
     run_OII_multinest
     run_OII_3
J1509p1506_source-30036
J1509p1506_source-30085
     run_OII_3
J1509p1506_source-30086
J1509p1506_source-30010
J1509p1506_source-30027
J1509p1506_source-30073
J1509p1506_source-30066
     run_cont
     run_OII_3
J1509p1506_source-30050
J1509p1506_source-30023
J1509p1506_source-30087
     run_OII_3
J1509p1506_source-30056
     run_cont
     run_HALPHA
J1509p1506_source-30029
J1509p1506_source-30007
J1509p1506_source-30004
J1509p1506_source-30047
     run_OII_multinest
     run_OII_3
J1509p1506_source-30062
     run_OII_3
J1509p1506_source-30057
     run_cont
     run_OII_multinest
     run_OII_3
J1509p1506_source-30026
J1509p1506_source-30092
     run_OII_3
J1509p1506_source-30032
     run_cont
J1509p1506_source-30093
     run_cont
     run_OII_multinest
     run_OII_3
J1509p1506_source-30080
J1509p1506_source-30019
J1509p1506_source-30082
     run_OII_multinest
     run_OII_3
J

     run_OII_3
J1107p1021_source-23148
J1107p1021_source-23030
J1107p1021_source-23008
J1107p1021_source-23110
     run_cont
     run_OII_3
J1107p1021_source-23106
     run_OII_multinest
     run_OII_3
J1107p1021_source-23038
J1107p1021_source-23111
     run_cont
     run_OII_3
J1107p1021_source-23126
     run_OII_3
J1107p1021_source-23152
     run_OII_multinest
     run_OII_3
J1107p1021_source-23140
J1107p1021_source-23086
     run_OII_multinest
     run_OII_3
J1107p1021_source-23044
J1107p1021_source-23138
     run_OII_3
J1107p1021_source-23033
J1107p1021_source-23061
J1107p1021_source-23003
J1107p1021_source-23118
     run_OII_multinest
     run_OII_3
J1107p1021_source-23089
     run_cont
     run_OII_multinest
J1107p1021_source-23141
J1107p1021_source-23142
     run_OII_multinest
     run_OII_3
J1107p1021_source-23015
J1107p1021_source-23102
     run_OII_multinest
     run_OII_3
J1107p1021_source-23029
J1107p1021_source-23146
J1107p1021_source-23113
J1107p1021_source-23091
     run

     run_OII_3
J0014p0912_source-12015
     run_OII_3
J0014p0912_source-12033
J0014p0912_source-12129
     run_OII_3
J0014p0912_source-12066
J0014p0912_source-12023
J0014p0912_source-12119
     run_OII_multinest
     run_OII_3
J0014p0912_source-12108
     run_cont
     run_OII_multinest
     run_OII_1
     run_OII_3
J0014p0912_source-12044
J0014p0912_source-12112
     run_OII_3
J0014p0912_source-12109
     run_OII_multinest
     run_OII_3
J0014p0912_source-12055
     run_OII_multinest
     run_OII_3
J0014p0912_source-12012
J0014p0912_source-12051
J0014p0912_source-12126
     run_OII_3
J0014p0912_source-12069
J0014p0912_source-12047
J0014p0912_source-12117
     run_OII_multinest
     run_OII_3
J0014p0912_source-12011
J0014p0912_source-12019
J0014p0912_source-12040
J0014p0912_source-12095
J0014p0912_source-12030
J0014p0912_source-12065
J0014p0912_source-12009
J0014p0912_source-12062
J0014p0912_source-12070
J0014p0912_source-12036
J0014p0912_source-12114
     run_OII_multinest
     run_OI

J0131p1303_source-16062
     run_cont
J0131p1303_source-16046
J0131p1303_source-16088
     run_cont
     run_OII_3
J0131p1303_source-16103
J0131p1303_source-16108
     run_OII_multinest
     run_OII_3
J0131p1303_source-16100
J0131p1303_source-16050
J0131p1303_source-16043
J0131p1303_source-16031
     run_OII_multinest
     run_OII_3
J0131p1303_source-16052
J0131p1303_source-16084
     run_cont
     run_OII_3
J0131p1303_source-16007
J0131p1303_source-16019
J0131p1303_source-16021
J1236p0725
J1236p0725_source-25007
J1236p0725_source-25005
     run_OII_3
J1236p0725_source-25088
     run_OII_3
J1236p0725_source-25079
J1236p0725_source-25008
J1236p0725_source-25051
J1236p0725_source-25078
J1236p0725_source-25076
     run_OII_multinest
     run_OII_3
J1236p0725_source-25060
     run_OII_3
J1236p0725_source-25085
     run_OII_multinest
     run_OII_3
J1236p0725_source-25093
     run_OII_3
J1236p0725_source-25090
     run_OII_3
J1236p0725_source-25021
J1236p0725_source-25082
J1236p0725_source-

     run_OIII5007_multinest
J1425p1209_substracted_cube.pdf
J1425p1209_source-29022
J1425p1209_source-29089
J1425p1209_source-29099
     run_cont
J1425p1209_source-29095
     run_OII_3
J1425p1209_source-29055
     run_cont
J1425p1209_source-29003
J1425p1209_source-29043
J1425p1209_source-29006
J1425p1209_source-29063
J1425p1209_source-29094
     run_OII_multinest
     run_OII_3
J1425p1209_source-29010
J1425p1209_source-29031
J1425p1209_source-29047
J1425p1209_source-29026
J1425p1209_source-29001
J1425p1209_source-29075
     run_cont
     run_OII_3
J1425p1209_source-29086
     run_OII_multinest
     run_OII_3
J1425p1209_source-29100
     run_OII_3
J1425p1209_source-29059
     run_cont
     run_OII_multinest
     run_OII_3
J1425p1209_source-29090
     run_OII_multinest
J1425p1209_source-29088
     run_OII_3
J1425p1209_source-29066
     run_cont
     run_OII_multinest
     run_OII_1
     run_OII_3
J1425p1209_source-29065
     run_OII_3
J1425p1209_source-29068
J1425p1209_source-29042
J1425

J0937p0656_source-21059
J0937p0656_source-21030
J0937p0656_source-21005
J0937p0656_source-21132
J0937p0656_source-21129
J0937p0656_source-21127
     run_OII_multinest
     run_OII_3
J0937p0656_source-21120
     run_cont
J0937p0656_source-21054
     run_OII_multinest
     run_OII_3
J0937p0656_source-21050
J0937p0656_source-21007
J0937p0656_source-21009
J0937p0656_source-21068
J0937p0656_source-21002
J0937p0656_source-21116
     run_cont
     run_OII_3
J0937p0656_source-21051
     run_OII_multinest
     run_OII_3
J0937p0656_source-21131
     run_OII_3
J0937p0656_source-21045
J0937p0656_source-21092
J0937p0656_source-21029
J0937p0656_source-21115
     run_cont
J0937p0656_source-21133
J0937p0656_source-21095
     run_cont
     run_OII_3
J0937p0656_source-21019
J0937p0656_source-21013
J0937p0656_source-21093
     run_OII_3
J0937p0656_source-21041
J0937p0656_source-21055
     run_OII_multinest
     run_OII_3
J0937p0656_source-21101
     run_OII_3
J0937p0656_source-21113
     run_cont
     ru

     run_OII_3
J0800p1849_source-19087
J0800p1849_source-19037
J0800p1849_source-19078
     run_OII_3
J0800p1849_source-19036
J0800p1849_source-19018
J0800p1849_source-19011
J0800p1849_source-19044
     run_OII_multinest
     run_OII_3
J0800p1849_source-19091
     run_OII_3
J0800p1849_source-19005
     run_cont
     run_OII_multinest
     run_OII_3
J0800p1849_source-19041
     run_OII_3
J0800p1849_source-19021
     run_OII_multinest
     run_OII_3
J0800p1849_source-19082
     run_cont
     run_OII_3
J0800p1849_continuum.pdf
J0800p1849_source-19003
     run_OII_3
J0800p1849_source-19017
J0800p1849_source-19096
J0800p1849_source-19025
J0800p1849_source-19068
     run_OII_multinest
     run_OII_3
J0800p1849_source-19067
     run_OII_multinest
     run_OII_3
J0800p1849_source-19024
J0800p1849_source-19060
     run_cont
     run_OII_multinest
     run_OII_3
J0800p1849_source-19031
J0800p1849_source-19052
J0800p1849_source-19102
     run_OII_multinest
     run_OII_3
J0800p1849_source-19056
 

     run_OII_3
J0014m0028_source-11038
J0014m0028_source-11027
J0014m0028_source-11073
J0014m0028_source-11082
J0014m0028_source-11151
     run_OII_3
J0014m0028_source-11129
     run_OII_multinest
     run_OII_3
J0014m0028_source-11026
J0014m0028_source-11085
J0014m0028_source-11011
J0014m0028_source-11050
J0014m0028_source-11158
J0014m0028_source-11002
J0014m0028_source-11092
J0014m0028_source-11130
     run_OII_multinest
     run_OII_3
J0014m0028_source-11118
     run_OII_3
J0014m0028_source-11104
     run_OII_2
     run_OII_multinest
     run_OII_3
J0014m0028_source-11041
J0014m0028_source-11121
     run_cont
     run_OII_2
     run_OII_multinest
     run_OII_3
J0014m0028_source-11150
     run_OII_3
J0014m0028_source-11108
     run_OII_2
     run_OII_1
     run_OII_3
J0014m0028_source-11025
J0014m0028_source-11162
     run_cont
J0014m0028_source-11155
     run_OII_3
J0014m0028_source-11154
     run_cont
     run_OII_3
J0014m0028_source-11142
     run_OII_2
     run_OII_multinest
   

In [53]:
runs2 = gt.match_results_with_catalogs(runs, dr2_path, fields_info_path, Abs_path, primary_tab_path,\
                                    output_path = output_path, media_path = media_path, dv_abs_match = 0.5e6, \
                                    export = False, b_max = 100, b_sep = 20, dv_primary = 0.5e6, \
                                       log_max_mass_satellite = 0, group_threshold = 4)

match with DR2
match with QSOs & fields
match with absorptions
computing the neighbours
     around each galaxy
     Closest
     around LOS
primary auto-identification
match with galpak results
Nb of row in the DR2 catalog:  2020
Nb of row in the galpak results:  1850
computing the alpha
calc score auto
calc SFR
calc virial mass, radius and halo parameters
object


In [54]:
cas = runs2[runs2["ID"] == 29002]
cas[["ID", "run_name", "chi2_at_p", "BIC", "score_auto"]]

ID                run_name  chi2_at_p      BIC  score_auto
2228  29002            run_OIII5007   1.876915  56014.0           0
2229  29002       run_OII_multinest   0.520300  15626.0           3
2230  29002               run_OII_3   1.157668  34596.0           0
2231  29002  run_OIII5007_multinest   0.498823  14977.0           3

In [55]:
R = gt.get_best_run2(runs2)

In [56]:
cas2 = R[R["ID"] == 15003]
cas2[["ID", "run_name", "chi2_at_p", "BIC"]]

ID                run_name  chi2_at_p      BIC
0  15003  run_OIII5007_multinest   0.553058  16592.0

In [57]:
len(R[R["run_name"] == "run_OII_multinest"])

454

We check the specific cases:

In [45]:
cases_id = [11083, 12103, 14127, 15003, 17068, 20077, 23086, 23120, 24038, 25070, 26056, 27094, 29002,\
            29084, 30070, 31005]

cases = R[R["ID"].isin(cases_id)]
cases[["ID", "run_name"]]

ID                run_name
0  11083                     NaN
0  12103       run_OII_multinest
0  14127       run_OII_multinest
0  15003  run_OIII5007_multinest
0  17068       run_OII_multinest
0  20077       run_OII_multinest
0  23086       run_OII_multinest
0  23120       run_OII_multinest
0  24038       run_OII_multinest
0  25070  run_OIII5007_multinest
0  26056     run_HBETA_multinest
0  27094       run_OII_multinest
0  29002  run_OIII5007_multinest
0  29084       run_OII_multinest
0  30070       run_OII_multinest
0  31005       run_OII_multinest

# We check that each absorption is affected to a unique galaxy

In [46]:
len(R["abs_id"].unique())

122

In [47]:
prim = R[R["primary_auto"] == 1]
isolated = R[R["isolated_auto"] == 1]
prim_with_abs = prim[prim["bool_absorption"] == 1]
isolated_with_abs = isolated[isolated["bool_absorption"] == 1]
print("Number of primary galaxies: ", len(prim))
print("Number of isolated galaxies: ", len(isolated))
print("Number of primaries associated with an abs: ", len(prim_with_abs))
print("Number of distinct absorptions associated with primaries: ", len(prim_with_abs["abs_id"].unique()))
print("Number of isolated associated with an abs: ", len(isolated_with_abs))
print("Number of distinct absorptions associated with isolated: ", len(isolated_with_abs["abs_id"].unique()))

Number of primary galaxies:  191
Number of isolated galaxies:  176
Number of primaries associated with an abs:  54
Number of distinct absorptions associated with primaries:  53
Number of isolated associated with an abs:  45
Number of distinct absorptions associated with isolated:  44


one absorption is associated with 2 primary / isolated galaxies

In [48]:
for i, k in isolated_with_abs.iterrows():
    print(k["ID"], k["field_id"], k["Z"], k["abs_id"], k["vel_absorption_dist"]/1e3)

11053 J0014m0028 0.65901 5 377.67478027257874
11083 J0014m0028 0.656 5 166.55136555554125
11094 J0014m0028 0.58391 104 350.15628874113565
11130 J0014m0028 1.1638 109 19.396868527594364
13063 J0015m0751 0.73143 9 39.82388276743445
13088 J0015m0751 0.81654 10 193.09080772237402
13128 J0015m0751 0.50687 8 119.37026737540741
15012 J0103p1332 0.76414 14 1.6993688596254493
15038 J0103p1332 1.04851 16 35.12318217631614
15040 J0103p1332 0.78817 15 45.26636933848445
16070 J0131p1303 1.0104 18 67.10436037605373
16101 J0131p1303 0.92665 17 231.84842209015517
17080 J0134p0051 1.06992 146 37.656546668459704
18049 J0145p1056 0.54994 25 88.97410911390426
18063 J0145p1056 0.76993 26 88.07810374420993
18092 J0145p1056 0.55609 150 256.2345167310263
19067 J0800p1849 0.73503 157 48.38065522785433
19073 J0800p1849 0.84331 29 95.95648600614862
19103 J0800p1849 0.99337 30 54.142123579670645
20052 J0838p0257 1.10063 164 215.5004030124154
20064 J0838p0257 1.33043 33 55.31629653753992
20066 J0838p0257 1.39663 1

## Now we manually inspect/remove the cases with mergers, psf issues etc:

- 11053: very low OII; ZCONF = 1, remove it.
- 11094: very low OII: but ZCONF = 3, maybe we can keep it.. at least for W-b
- 12103: OII faint, but src is large. we should keep it
- 17068: OII ok, src very extended. We should keep it, maybe rerun. smooth camel
- 18082: OII ok, but src very compact. We should keep it
- 21117: OII ok, but close to QSO psf. Multinest run failed. Rerun, we should keep it.
- 23120: otehr src nearby, make the runs fail. remove it, or keep if at least for W-b
- 23139: very irregular source. OII faint, maybe remove it
- 24052: ZCONF = 2, OII run failed. maybe we can keep it, but need to rerun.
- 26021: ZCONF = 1, very compact. Remove
- 26056: merger or very extended strange source... Keep it for W-b and fc?
- 28002: close to psf. No run converged (multinest failed). Rerun, we should keep it at least for W-b
- 28074: OII OK. But very compact. We should keep it at least for W - b and fc
- 28081: faint src, but ZCONF = 3. Bad runs, no convergence. Only keep it for W-b and fc.
- 29002: completely mixed src with the psf. keep it only for W-b and fc but not usable with galpak.
- 29084: ZFONF = 3; we should keep it for W-b and fc but not usable with galpak: run completely fail because of psf.
- 30017: at the same z as another src, so not good for galpak and not isolated. Remove.
- 30020: no OII visible... ZCONF = 2 but should be 1?? remove
- 30070: OII faint, not isolated. 130kpc... remove.
- 31002: looks OK. ZCONF= 3. Keep it.
- 31003: low OII. But seems ok. Multinest run failed. Keep it at least for W - b.
- 31005: 2 srcs merged. Keep it only for w-b and fc. Not for galpak.
- 32053: rather ok. ZCONF = 3. convergence not very good. But Keep it.
- 32074: ZCONF = 3, quite faint. But ok for W-b and fc.

In [12]:
#bad_ids = [11053, 11094, 12103, 17068, 18082, 21117, 23120, 23139, 24052, 26021, 26056, 28002, 28074, 28081, 29002,\
#           29084, 30017, 30020, 30070, 31002, 31003, 31005, 32053, 32074]

In [58]:
R.to_csv("results_7_conservative.csv")

In [14]:
#gt.build_catalog(prim, output_path, output_path, file_name = "primary_catalog_conservative")

In [15]:
#ids = R[R["ID"].isin([23102, 28056, 28070])]

In [16]:
#gt.build_velocity_map_on_ids(input_path, output_path, ids, snr_min = 1, commonw=True, dv=500., \
#                             dxy=15, deltav=2000., initw=50., wmin=30., wmax=250., dfit=100., degcont=0,\
#                             sclip=10, xyclip=3, nclip=3, wsmooth=0, ssmooth=2)

In [17]:
run_dir = "/muse/MG2QSO/private/analysis/galpak_dr2/"
output_dir = "/muse/MG2QSO/private/analysis/galpak_dr2/"
#gt.build_catalog(isol_9, output_path, output_path, file_name = "isolated_catalog_2023_10_18")

field_id = r["field_id"]
src_id = r["ID"] 
run_name = r["run_name"]
z_src = r["Z"]
REW2796 =r["REW_2796"]
B_KPC = r["B_KPC"]
incl = r["inclination"]
alpha = r["alpha"]
radius = r["radius"]
snr_eff = r["snr_eff"]
#score = r["galpak_score"]
#score_auto = r["score_auto"]
#primary = r["primary"]
primary_auto = r["primary_auto"]
isolated_auto = r["isolated_auto"]
logMass = r["sed_logMass"]

#print(output_dir, run_name, src_id, field_id)


title = field_id+" - " +str(src_id)+" - z = "+str(np.round(z_src, 3)) +" - B = "+str(B_KPC)+\
    " - REW2796 = "+str(REW2796)+" - log(M) = "+str(np.round(logMass,2)) + " - SNReff = " +str(np.round(snr_eff,1))+\
    " - incl = "+ str(np.round(incl,1))+" - alpha = "+str(np.round(alpha,1)) + " radius = " + str(np.round(radius,2))+  "\n "+\
    "run = " + str(run_name) +  " isolated = "+str(isolated_auto) 

fig = plt.figure(figsize = (18,5))
fig.suptitle(title)

#print(type(run_name))
if type(run_name) is str:
    run_path = run_dir + field_id +"/"+field_id +"_source-"+str(src_id)+"/"+str(run_name)+"/"
    print(run_path)

    # For the measured flux:
    img_run = plt.imread(run_path + "run_images.png")
    img_measured = img_run[10:320, 140:440]

    # For th model:
    img_model_conv = plt.imread(run_path + "run_obs_maps.png")
    img_model_conv = img_model_conv[50:320, 430:730]

    plt.subplot(131)
    plt.imshow(img_measured)
    plt.axis("off")

    plt.subplot(132)
    plt.imshow(img_model_conv)
    plt.axis("off")

    # For the velmap:
    try:
        vel = "camel_OII/camel_"+ str(src_id) +"_"+"o2_ssmooth" +"_vel_common.fits"
        snr = "camel_OII/camel_"+ str(src_id) +"_"+"o2_ssmooth" +"_snr_common.fits"
        velmap_path = run_dir + field_id +"/"+field_id +"_source-"+str(src_id)+"/"+vel
        snr_path = run_dir + field_id +"/"+field_id +"_source-"+str(src_id)+"/"+snr
        hdul_vel = fits.open(velmap_path)
        hdul_snr = fits.open(snr_path)
        img_vel = hdul_vel[0].data
        img_snr = hdul_snr[0].data
        m = np.where(img_snr>4, 1, 0)

        kpc_per_arcsec = cosmo.kpc_proper_per_arcmin(z_src).value/60
        extent_arcsec = np.array([-0.2*15, 0.2*15,-0.2*15, 0.2*15])
        extent_kpc = extent_arcsec*kpc_per_arcsec

        plt.subplot(133)
        #plt.title("velocity map (snr>4 mask)")
        plt.imshow(img_vel*m, vmin = -150, vmax = 150, cmap = "bwr", extent = extent_kpc)
        cbar = plt.colorbar(label = "Dv [km/s]")
        plt.xlabel("x [kpc]", size = 12)
        plt.ylabel("y [kpc]", size = 12)
        plt.gca().invert_yaxis()
        #fig.savefig(pdf, format='pdf')
    except:
        print(src_id, ": no vel. map")
        plt.subplot(133)
        plt.scatter(0,0, c = "white")

else:
    print(src_id ,": no run" , run_name, type(run_name))
    plt.subplot(131)
    plt.scatter(0,0, c = "white")
    plt.axis("off")

from mpl_toolkits.axes_grid1 import make_axes_locatable

field_id = r["field_id"]
src_id = r["ID"] 
run_name = r["run_name"]
z_src = r["Z"]
REW2796 =r["REW_2796"]
B_KPC = r["B_KPC"]
incl = r["inclination"]
alpha = r["alpha"]
radius = r["radius"]
snr_eff = r["snr_eff"]
#score = r["galpak_score"]
#score_auto = r["score_auto"]
#primary = r["primary"]
primary_auto = r["primary_auto"]
isolated_auto = r["isolated_auto"]
logMass = r["sed_logMass"]

#print(output_dir, run_name, src_id, field_id)


title = field_id+" - " +str(src_id)+" - z = "+str(np.round(z_src, 3)) +" - B = "+str(B_KPC)+\
    " - REW2796 = "+str(REW2796)+" - log(M) = "+str(np.round(logMass,2)) + " - SNReff = " +str(np.round(snr_eff,1))+\
    " - incl = "+ str(np.round(incl,1))+" - alpha = "+str(np.round(alpha,1)) + " radius = " + str(np.round(radius,2))+  "\n "+\
    "run = " + str(run_name) +  " isolated = "+str(isolated_auto) 

fig = plt.figure(figsize=(14, 14), dpi = 250)
gs = fig.add_gridspec(3, 3, left=0.1, right=0.9, bottom=0.1, top=0.9,
                      wspace=0.2, hspace=0.0)
# Create the Axes.
ax_meas = fig.add_subplot(gs[0:1, 0:1])
ax_model = fig.add_subplot(gs[0:1, 1:2])
ax_velmap = fig.add_subplot(gs[0:1, 2:3])
ax_conv = fig.add_subplot(gs[1:3, 0:3])

fig.suptitle(title)

#print(type(run_name))
if type(run_name) is str:
    run_path = run_dir + field_id +"/"+field_id +"_source-"+str(src_id)+"/"+str(run_name)+"/"
    print(run_path)

    # For the measured flux:
    img_run = plt.imread(run_path + "run_images.png")
    img_measured = img_run[10:320, 140:440]

    # For the model:
    img_model = plt.imread(run_path + "run_obs_maps.png")
    img_model = img_model[50:320, 430:730]
    
    # For the convergence:
    img_conv = plt.imread(run_path + "run_mcmc.png")

    ax_meas.imshow(img_measured)
    ax_meas.axis("off")

    ax_model.imshow(img_model)
    ax_model.axis("off")
    
    ax_conv.imshow(img_conv)
    ax_conv.axis("off")

    # For the velmap:
    #try:
    vel = "camel_OII/camel_"+ str(src_id) +"_"+"o2_ssmooth" +"_vel_common.fits"
    snr = "camel_OII/camel_"+ str(src_id) +"_"+"o2_ssmooth" +"_snr_common.fits"
    velmap_path = run_dir + field_id +"/"+field_id +"_source-"+str(src_id)+"/"+vel
    snr_path = run_dir + field_id +"/"+field_id +"_source-"+str(src_id)+"/"+snr
    hdul_vel = fits.open(velmap_path)
    hdul_snr = fits.open(snr_path)
    img_vel = hdul_vel[0].data
    img_snr = hdul_snr[0].data
    m = np.where(img_snr>4, 1, 0)

    kpc_per_arcsec = cosmo.kpc_proper_per_arcmin(z_src).value/60
    extent_arcsec = np.array([-0.2*15, 0.2*15,-0.2*15, 0.2*15])
    extent_kpc = extent_arcsec*kpc_per_arcsec

    #plt.subplot(133)
    #plt.title("velocity map (snr>4 mask)")
    divider = make_axes_locatable(ax_velmap)
    cax = divider.append_axes('right', size='5%', pad=0.05)
    im = ax_velmap.imshow(img_vel*m, vmin = -150, vmax = 150, cmap = "bwr", extent = extent_kpc)
    fig.colorbar(im, cax=cax, orientation='vertical')
    #plt.colorbar(ax_velmap, label = "Dv [km/s]")
    ax_velmap.set_xlabel("x [kpc]", size = 12)
    ax_velmap.set_ylabel("y [kpc]", size = 12)
    ax_velmap.invert_yaxis()
    #fig.savefig(pdf, format='pdf')
    #except:
     #   print(src_id, ": no vel. map")
        #plt.subplot(133)
     #   ax_velmap.scatter(0,0, c = "white")

else:
    print(src_id ,": no run" , run_name, type(run_name))
    #plt.subplot(131)
    ax_meas.scatter(0,0, c = "white")
    ax_meas.axis("off")